### PR TITLE
Improve REST API error mapping

### LINKED ISSUE #17441: Improve error mapping of REST API
**Description**

Several common errors may be thrown in the REST API which are currently incorrectly mapped to the internal server error code (500):

- [x] `BrokerClient` `TimeoutException`, which should be mapped to gateway timeout (502).
- [x] `MessagingException$ConnectionClosed`, which should be mapped to gateway timeout (502).
- [x] `BrokerClient` `ConnectException`, which should be mapped to service unavailable (503).
- [x] `BrokerClient` `ConnectTimeoutException`, which should be mapped to service unavailable (503).
- [x] `BrokerClient` `PartitionNotFoundException`, which should be mapped to service unavailable (503).
- [x] `MsgPackException`, `JsonParseException`, `IllegalArgumentException`, which should all map to 400 (or some variant), indicating the request is the issue.
- [x] `RequestRetriesExhaustedException` should be mapped to 429.

Possibly I missed some. But the above are important to allow us to continue computing the SLI/SLO/SLA correctly, so I would urge the priority to be increased. While it's not a bug per se (since, well, this isn't "breaking" anything), happy to have it treated as one.

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Improve the error mapping of the REST API.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #17441

### CODE REVIEW

**Comment by koevskinikola** [comment]
@tmetzke @nicpuppa I added you both as reviewers to merge the PR faster. I'll merge after the first approval 😃

**Review by nicpuppa — APPROVED** [review]
👍

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:118]: 🔧 I guess this should be gateway timeout (504), not bad gateway (502), right?

Comment (tmetzke) [reply zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:118]:
```suggestion
        yield createProblemDetail(HttpStatus.GATEWAY_TIMEOUT, teMsg, te.getClass().getName());
```

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:124]:
```suggestion
        yield createProblemDetail(HttpStatus.GATEWAY_TIMEOUT, ccMsg, cc.getClass().getName());
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:269]:
```suggestion
        .isEqualTo(HttpStatus.GATEWAY_TIMEOUT)
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:247]:
```suggestion
  public void shouldReturnGatewayTimeoutOnTimeoutException() {
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:277]:
```suggestion
  public void shouldReturnGatewayTimeoutOnConnectionClosed() {
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:286]:
```suggestion
            HttpStatus.GATEWAY_TIMEOUT,
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:255]:
```suggestion
            HttpStatus.GATEWAY_TIMEOUT,
```

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/ErrorMapperTest.java:301]:
```suggestion
        .isEqualTo(HttpStatus.GATEWAY_TIMEOUT)
```

**Pull request merged**
